### ISSUE TITLE
Implement upgrades (only) as classes

### ISSUE THREAD

**Issue by juho-jaakkola** [Elgg/Elgg#6891]
Just writing down some thoughts.

We did some improvement for running batch upgrades in 1.9 through the ```ElggUpgrade``` object and related views and js.

All upgrades however still require their own view and action, which is unnecessary duplication.

Goal: Creating an upgrade means creating a class. Nothing else is needed. All upgrade logic is encapsulated inside the class. Elgg core will take care of listing those classes, running the upgrades and keeping track of which upgrades have been ran.

Lets add:
 - Interface that defines the structure of an upgrade class
 - Single view that can display any upgrade
  - The view gets all necessary information from the class and displays them to the user
 - Single action that takes care of all upgrades
  - The action gets all necessary information as parameters and it passes them to the class
 
Each class would have an unique id which would be used:
 - To tell the action which upgrade to run
 - To keep track which upgrades have been ran
 - To define dependencies to other upgrades (The UI would tell user that they need to run upgrade A before they can run upgrade B)

**Comment by ewinslow** [comment]
Seems like what we've been proposing as BatchUpgrades (i.e. many repetetive steps, able to show progress meter in a UI) is a good level of generalization. Anything that has only a single step can just go straight from 0-100% finished. I propose we just have a single `Upgrade` class that does this and model everything after that use-case.

#### Developer experience

Example upgrade for annotation->entity comments:

```php
namespace Elgg\Comments\Upgrades;

use Elgg\Storage\Annotation;
use Elgg\Storage\Entity;
use Elgg\Storage\Repository;
use Elgg\Storage\Upgrade;

class ConvertCommentAnnotationsToEntities implements Upgrade<Annotation> {
  public function __construct(Repository<Entity> $entities, Repository<Annotation> $annotations) {
    $this->entities = $entities;
    $this->annotations = $annotations;
  }

  /** @override */
  public function getRemainingItems(): Collection<Annotation> {
    return $this->annotations->where(['type' => 'generic_comment']);
  }

  /** @override */
  public function process(Annotation $comment) {
    $this->entities->create([
      'type' => 'object',
      'subtype' => 'comment',
      'description' => $comment->value,
      'owner_guid' => $comment->owner_guid,
      'container_guid' => $comment->entity_guid,
      'time_created' => $comment->time_created,
    ]);

    $this->annotations->delete($comment->id);
  }
}
```


#### Admin experience

 * A single page lists all the pending upgrades
 * There is a button to run all upgrades
 * Upgrades are run one at a time, each with a live-updating progress meter
   * This is accomplished by saving the initial number of remaining items client-side, then re-fetching that number periodically to check progress. If it started at 100 items remaining and there are 10 items left, then we are 90% complete.
   * This implies that if you refresh the page, the "initial" count should change. That's ok, and by design.

#### Implementation details

* Upgrades are identified by their fully qualified class name. No generator script necessary.
* Available upgrades are automatically detected by Elgg based on which classes implement `Elgg\Storage\Upgrade`.

```php
interface Upgrade<ITEM> {
  /**
   * Return all items left to be upgraded.
   */
  public function getRemainingItems(): Collection<ITEM>;

  /**
   * Expected to throw an exception if the migration fails for any reason.
   */ 
  public function process(ITEM $item): void;
}

final class UpgradeProcessor {
  private $upgrade;
  
  public function __construct(Upgrade $upgrade) {
    $this->upgrade = $upgrade;
  }

  public function getNumRemaining() {
    return count($this->upgrade->getRemainingItems());
  }

  public function execute() {
    foreach ($this->upgrade->getRemainingItems() as $item) {
      $this->upgrade->process($item);
    }
  }
}
```

**Comment by ewinslow** [comment]
Forgive the pseudo-cody nature of the above proposal/examples. I've been liking generics lately and have been dreaming about OO APIs for Elgg's data model for a long time...

**Comment by mrclay** [comment]
Most but not all upgrades will individually process a countable set of items, but as long as we can cram unusual upgrades into this interface SGTM.

**Comment by juho-jaakkola** [comment]
Fixed in #10118
